pilot/pkg/proxy/envoy/v2/ads.go

```diff
@@ -95,6 +97,18 @@ type XdsConnection struct {
 	LDSWatch bool
 	// CDSWatch is set if the remote server is watching Clusters
 	CDSWatch bool
+
+	// Envoy may request different versions of configuration (XDS v2 vs v3). While internally Pilot will
+	// only generate one version or the other, because the protos are wire compatible we can cast to the
+	// requested version. This struct keeps track of the types requested for each resource type.
+	// For example, if Envoy requests Clusters v3, we would track that here. Pilot would generate a v2
+	// cluster response, but change the TypeUrl in the response to be v3.
```

Comment:
Note that this only works in the absence of using v2 deprecated fields or v3 new fields.

Reply:
@howardjohn We must make sure not use any deprecated fields of v2 when v3 is requested.